### PR TITLE
Add various servers and rework caching solution

### PR DESCRIPTION
**Pull request by matthewh86**

### CODE REVIEW

**Review by sirsquidness — COMMENTED** [review]
Thank you for your extensive work here! Lots of overdue cleanups and upgrades :)

On the caching system, I want to refine the design a bit more.

It seems to me that your use case you're solving is ... you build your images on the same server that you use them on, and you don't want to re-download the 15GB-or-so CSGO every time the base image is updated. Does that sound correct?

I use these docker images with Kubernetes and local private dockerhubs (at least I did back before the pandemic when live events still existed). So I build, push to dockerhub over network, and 1 or more kube servers pull the image. So image layering efficiency is important to me - both in terms of total size, and how long it takes to push an update. (this is also why I use timestamp tags). 15GB images take a lot of time to pull over the network and extract on disk, and when Valve drops a patch in the middle of a tournament, time matters to me.

In the solution in this PR, every time there's a game update, the `ADD cache/ .`  layer in the dockerfile would be totally rebuilt. In the case of CSGO, this means a brand new 15GB layer every time.

Here's some pseudo code that I think would solve both your use case and mine at the same time:

```

# new script: update_cache.sh or something
steamID= $(cat csgo/steamid)
if steamID != ""
   # This could use a docker container, or could use steamcmd on the local system, doesn't matter
   steamcmd_download $steamID csgo/.cache
endif

# inside build_docker.sh
# Makes sure .cache exists, even if empty, so that the docker file can use `COPY` on it - the dockerfile doesn't know if the
# directory exists or not, except that the build fails if it doesn't, so we have to make sure it exists even fi the cache isn't populated
mkdir -p $game/.cache

# Dockerfile for csgo
# Add the cache - this could be empty, or it could be a slightly out of date copy of the game, or a fully up to date copy
# Updating the cache will invalidate the layer cache of this, so updating the cache causes the entire image to be effectively rebuilt
COPY--chown steam .cache/ /steam/csgo
# Run the steamcmd update, so even if the cache was out of date or empty, there will be an up to date copy of the game
# This layer will get cached as is by docker, and guarantees useful cached layer exists for updates to be based on
RUN steamcmd update $csgoappid -dir /steam/csgo

# So if there's a game update, and you don't want to distribute a brand new 15GB layer and want to do an incremental update,
# set the build arg CACHE_DATE to `$(date)` and this last layer will contain just the diff between the existing docker image and
# the update. If there are no updates to make, this layer is effectively a no-op
ARG CACHE_DATE
RUN steamcmd update $csgoappid -dir /steam/csgo

```

so as a use, I'd do something like 

```
./update_cache csgo
./docker_build csgo
```

This slightly different approach has benefits for both of our use cases!

For my own use case, it allows me to make the most efficient use of docker image layers - small CSGO update is really fast to build and distribute. I always have a steam cache running, and rarely update the base image, so for my own workflows this means I can opt out of using the cache and just use the networked steam cache instead.

In general, buliding the cache before building the docker image means...
* you get to pick when you update the cache
* which means successive docker builds become idempotent unless you update the cache - running `docker_build.sh csgo` twice in a row will result in building the image the first time, and using all cached layers the second time
* and also means there's no post-build step on every build to copy the game back out of the built image to the cache (big time saver in the case of CSGO!) (which also helps keep successive rebuilds idempotent)
* Small updates to the image can rely on the `CACHE_DATE` build arg instead of the local cache, which can be a lot faster.


Would love to hear your thoughts! Does it solve your use case? Does it align with the vision you had for the PR?

This PR is much appreciated and I look forward to merging it once we're converged on a caching flow that works for both of us

File: css/Dockerfile [hunk 1]
```diff
@@ -4,18 +4,22 @@ FROM steamcmd
...
+RUN mkdir -p /steam/${GAME}/.cache/ \
+  && touch /steam/${GAME}/.cache/empty
+WORKDIR /steam/${GAME}/
+COPY .cache .
 
-WORKDIR /steam/steamcmd_linux
-RUN ./steamcmd.sh +login anonymous +force_install_dir ../css +app_update 232330 +quit
+RUN chown -R steam /steam/${GAME}/ \
+  && mv .cache/* . \
+  && rm -fr .cache
```
[the remark this answers is not in the document]
Comment (matthewh86) [reply css/Dockerfile:17]: Having a multi line RUN command will only create one layer. This is a bit of a moot point though as the caching "solution" I've added turns out to be quite slow locally.

I'm thinking that mounting a volume and running steamcmd update on the ENTRYPOINT rather than in the build will reduce image size, build time, and data usage. Especially with subsequent builds that change the underlying image layers.

[the remark this answers is not in the document]
Comment (sirsquidness) [reply css/Dockerfile:17]: If you're meaning that you'd keep the CSGO or whatever game files outside of the container and mount them in using a volume, that's not gonna fly here. The game files must be inside the image, and must be installed/updated during image build - anything else isn't really 

Perhaps a solution that would work well for you - you can use a regular LANCache type thing so that the game install goes super quick. No need for complicated or hacky solutions here - they'll still install super fast. You could even have the cache run on your local docker installation if you wanted to avoid the cost of going over gigabit network. The steamcmd updates support using steamcache no problems.

**Comment by matthewh86** [comment]
Hey @sirsquidness!

TLDR: I want to change the technical direction of this repo to not bake in the game data into the images.

Sorry for the long-standing PR, and thanks again for getting this repo up and running.

I generally only get to work on this in a lead up to a LAN party, and recently I've been doing some development on this to streamline getting the servers up and working and trying to do a bit of standardization. I think unfortunately the technical direction I want to take this next conflicts with on of the previous comments about having the game files baked in to the image (https://github.com/OpenSourceLAN/gameservers-docker/pull/25#discussion_r703457959). After the introduction of the `CACHE_DATE` ARG I had no problems, but I'm now coming across a totally different problem that necessitates me to change how I bake these images.

I'm wanting to host the various gameservers within AWS but the costs of merely keeping the images inside ECR ($1/10GB/month) far outweighs the cost of mounting the game files into the image via EBS (~$0.015/10GB/month).

https://calculator.aws/#/estimate?id=d281cac8f0b60b5ff3ba11915179a7bcacf41663

Untitled-2024-09-18-2130

I shoved a lot of the gameservers I haven't really worked on into a legacy folder, but I can move them back for this PR.

LMK if you want me to make any changes to this PR so you can benefit from them before I go off in a different direction!

EDIT: Hmm I just thought that an alternative to the above would be if I self-hosted a docker repository within AWS. This way we can maintain baking the game data into the images... I'll have to ruminate some more.

File: trackmania-forever/start.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+sed -i "s/%%SERVER_COMMENT%%/$SERVER_COMMENT/" /tm/GameData/Config/dedicated_cfg.txt
+sed -i "s/%%SERVER_PASSWORD%%/$SERVER_PASSWORD/" /tm/GameData/Config/dedicated_cfg.txt
+sed -i "s/%%MAX_PLAYERS%%/$MAX_PLAYERS/" /tm/GameData/Config/dedicated_cfg.txt
+
+sed -i "s/%%SUPER_ADMIN_PASSWORD%%/$SUPER_ADMIN_PASSWORD/" /tm/GameData/Config/dedicated_cfg.txt
+sed -i "s/%%ADMIN_PASSWORD%%/$ADMIN_PASSWORD/" /tm/GameData/Config/dedicated_cfg.txt
+
+sed -i "s/%%MASTER_SERVER_LOGIN%%/$MASTER_SERVER_LOGIN/" /tm/GameData/Config/dedicated_cfg.txt
+sed -i "s/%%MASTER_SERVER_PASSWORD%%/$MASTER_SERVER_PASSWORD/" /tm/GameData/Config/dedicated_cfg.txt
+sed -i "s/%%MASTER_SERVER_KEY%%/$MASTER_SERVER_KEY/" /tm/GameData/Config/dedicated_cfg.txt
```
[the remark this answers is not in the document]
Comment (matthewh86) [reply trackmania-forever/start.sh:26]: No idea! We didn't get to play this and I haven't run it recently.

**Comment by sirsquidness** [comment]
hello, my apologies in return for not following up on this PR with you. Every now and again I see the commit update emails and I put them aside to remind me to come back to you and then life gets in the way and I forget about the notifications

Since I'm really bad at replying to notifications, maybe hit me on Discord (same username as here) and let's chat there to align on solutions and project direction if you're up for that?
